pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,709 @@
+		let max_rewards = accumulated_rps_now
+			.checked_sub(user_accumulated_rps)
+			.ok_or(Error::<T>::Overflow)?
+			.checked_mul(user_shares)
+			.ok_or(Error::<T>::Overflow)?;
+
+		let claimable_rewards = loyalty_multiplier
+			.checked_mul_int(max_rewards)
+			.ok_or(Error::<T>::Overflow)?;
+
+		let unclaimable_rewards = max_rewards.checked_sub(claimable_rewards).ok_or(Error::<T>::Overflow)?;
+
+		let user_rewards = claimable_rewards
+			.checked_sub(user_accumulated_claimed_rewards)
+			.ok_or(Error::<T>::Overflow)?;
+
+		Ok((user_rewards, unclaimable_rewards))
+	}
+
+	pub fn claim_from_global_pool(
+		pool: &mut GlobalPool<T::BlockNumber, Balance, T::CurrencyId, T::AccountId, T::BlockNumber>,
+		shares: Balance,
+	) -> Result<Balance, Error<T>> {
+		let reward = pool
+			.accumulated_rps
+			.checked_sub(pool.accumulated_rps_start)
+			.ok_or(Error::<T>::Overflow)?
+			.checked_mul(shares)
+			.ok_or(Error::<T>::Overflow)?
+			.min(pool.accumulated_rewards);
+
+		pool.accumulated_rps_start = pool.accumulated_rps;
+
+		pool.paid_accumulated_rewards = pool
+			.paid_accumulated_rewards
+			.checked_add(reward)
+			.ok_or(Error::<T>::Overflow)?;
+
+		pool.accumulated_rewards = pool
+			.accumulated_rewards
+			.checked_sub(reward)
+			.ok_or(Error::<T>::Overflow)?;
+
+		return Ok(reward);
+	}
+
+	pub fn update_pool(
+		pool_id: PoolId,
+		pool: &mut LiquidityPool<T::BlockNumber, Balance>,
+		rewards: Balance,
+		period_now: T::BlockNumber,
+		global_pool_id: PoolId,
+		reward_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		if pool.updated_at == period_now {
+			return Ok(());
+		}
+
+		if pool.total_shares.is_zero() {
+			return Ok(());
+		}
+
+		pool.accumulated_rps = Self::get_accumulated_rps(pool.accumulated_rps, pool.total_shares, rewards)?;
+		pool.updated_at = period_now;
+
+		let global_pool_balance =
+			T::MultiCurrency::free_balance(reward_currency, &Self::pool_account_id(global_pool_id)?);
+
+		ensure!(
+			global_pool_balance >= rewards,
+			Error::<T>::InsufficientBalancdInGlobalPool
+		);
+
+		let global_pool_account = Self::pool_account_id(global_pool_id)?;
+		let pool_account = Self::pool_account_id(pool_id)?;
+		T::MultiCurrency::transfer(reward_currency, &global_pool_account, &pool_account, rewards)
+	}
+
+	pub fn validate_pool_id(pool_id: PoolId) -> Result<(), Error<T>> {
+		if pool_id.le(&Zero::zero()) {
+			return Err(Error::<T>::InvalidPoolId);
+		}
+
+		Ok(())
+	}
+
+	pub fn validate_create_farm_data(
+		total_rewads: Balance,
+		planned_yielding_periods: T::BlockNumber,
+		blocks_per_period: T::BlockNumber,
+		yield_per_period: Permill,
+	) -> DispatchResult {
+		ensure!(
+			total_rewads >= T::MinTotalFarmRewards::get(),
+			Error::<T>::InvalidTotalRewards
+		);
+
+		ensure!(
+			planned_yielding_periods >= T::MinPlannedYieldingPeriods::get(),
+			Error::<T>::InvalidPlannedYieldingPeriods
+		);
+
+		ensure!(!Zero::is_zero(&blocks_per_period), Error::<T>::InvalidBlocksPerPeriod);
```

Comment:
Looks more readable.

```suggestion
		ensure!(!blocks_per_period.is_zero(), Error::<T>::InvalidBlocksPerPeriod);
```